src/avx512bw/avx512bw-from-utf8.inl.cpp

```diff
@@ -0,0 +1,462 @@
+/*
+validate_utf8_structure only checks if leading bytes
+are follewed by proper number of continuation bytes
+and then there's another leading byte.
+
+We don't handle end-of-stream here.
+*/
+bool validate_utf8_structure(__m512i input) {
+    /* 1. bitmask for various character byte classes.
+
+        leading: 111010011100011010001
+                 abccdddefgggghiijjjjk   a-k -- 11 UTF8- chars characters
+
+        ascii:   110000011000010000001
+                 ab     ef    h      k
+
+        2 bytes: 001000000000001000000
+                   cc          ii
+
+        3 bytes: 000010000000000000000
+                     ddd              
+
+        4 bytes: 000000000100010010000
+                          gggg   jjjj
+    */
+    uint64_t leading;
+    uint64_t ascii;
+    uint64_t _2bytes;
+    uint64_t _3bytes;
+    uint64_t _4bytes;
+
+    // we can valide 60 - 4 leading bytes
+    constexpr uint64_t mask  = 0x0ffffffffffffffflu;
+
+    {
+        const __m512i t0 = _mm512_and_si512(input, v_c0);
+        leading = _mm512_cmpneq_epu8_mask(t0, v_80);
+    }
+    {
+        ascii = _mm512_testn_epi8_mask(input, v_80) & mask;
+    }
+    {
+        const __m512i t0 = _mm512_and_si512(input, v_e0);
+        _2bytes = _mm512_cmpeq_epi8_mask(t0, v_c0) & mask;
+    }
+    {
+        const __m512i t0 = _mm512_and_si512(input, v_f0);
+        _3bytes = _mm512_cmpeq_epi8_mask(t0, v_e0) & mask;
+    }
+    {
+        const __m512i t0 = _mm512_and_si512(input, v_f8);
+        _4bytes = _mm512_cmpeq_epi8_mask(t0, v_f0) & mask;
+    }
+
```

Comment:
Generally, I'd look at what the compiler does with this function, and consider maybe using things like _kshiftri_mask64 if the compilers are too dumb.

Reply:
> > Please be aware that my next goal is to port your UTF-8 validation that relies on pshufb.
> 
> My suspicion is that it would be much faster than what you have here if implemented properly. :-) And interesting in its own right!!!

I have similar feelings when finished this. There are some edge cases that might be costly. On the other hand, I benchmarked this algorithm --- the numbers I sent you last year came from this implementation. Also, I wanted to finish it, as it's weird enough. :) It may end up like the vectorized surrogate pairs handling --- doable, but we learned does not meet expectations.

My only minor concern regarding your elegant solution is the lack of chap byte-level shifts in AVX512BW. In AVX512VBMI we have reg-wide shuffles. But we will see, I'm optimistic.